File: Data Structures & Algorithms/house-robber-ii/submission-1.cpp
class Solution {
public:
    int solve(vector<int>& nums, int start, int end){
        int prev = nums[start];
        int prev2 = 0;
        for(int i  = start+1; i < end; i++){
            int take = nums[i];
            if(i > start+1) take += prev2;
            int notake = prev;
            int curr = max(take, notake);
            prev2 = prev;
            prev = curr;
        }
        return prev;
    }
    int rob(vector<int>& nums) {
        int a1 = solve(nums, 0, nums.size()-1);
        int a2 = solve(nums, 1, nums.size());
        return max(a1,a2);
    }
};

File: Data Structures & Algorithms/house-robber-iii/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rob(self, root: Optional[TreeNode]) -> int:
        cache = {}

        def solve(node):
            if not node:
                return 0
            if node in cache:
                return cache[node]
            take = node.val
            if node.left:
                take += solve(node.left.left) + solve(node.left.right)
            if node.right:
                take += solve(node.right.left) + solve(node.right.right)
            notake = solve(node.left) + solve(node.right)
            cache[node] = max(take, notake)
            return max(take, notake)
        solve(root)
        return cache[root]
    
            

